kentik_api_library/kentik_api/public/query_object.py

```diff
@@ -132,44 +132,146 @@ class AggregateFunctionType(Enum):
     notEquals = "notEquals"
 
 
+AggregateType = TypeVar("AggregateType", bound="Aggregate")
+
+
 @dataclass
 class Aggregate:
     name: str
     column: str
     fn: AggregateFunctionType
+    value: Optional[str] = None
+    label: Optional[str] = None
+    origLabel: Optional[str] = None
+    unit: Optional[str] = None
+    group: Optional[str] = None
     sample_rate: int = 1
     rank: Optional[int] = None  # valid: number 5..99; only used when fn == percentile
     raw: Optional[bool] = None  # required for topxchart queries
 
+    @classmethod
+    def from_dict(cls: Type[AggregateType], data: Dict) -> AggregateType:
```

Comment:
In requests_payload package we use dacite library to convert dict to a dataclass. Custom from_dict methods introduced here do not use dacite in order to check for mandatory fields, right?

Reply:
I was not really keen on writing all of those custom `from_dict` classes. However,  I've quickly tested using `kentik_api.requests_payload.conversions.from_dict` like this:
```
from kentik_api.requests_payload.conversions import from_dict

...

@dataclass
class QueryObject:
    queries: List[QueryArrayItem]
    imageType: Optional[ImageType] = None  # used in QueryChart
    version: int = 4

    @classmethod
    def from_dict(cls: Type[QueryObjectType], data: Dict) -> QueryObjectType:
        """
        Construct QueryObject based on a dictionary. The dictionary must contain data for all mandatory query attributes
        :param data: dictionary
        :return: instance of QueryObject
        """
        # verify that all QueryObject data field without built-in default value are provided
        missing = [field_name for field_name in mandatory_dataclass_attributes(cls) if field_name not in data]
        if missing:
            raise RuntimeError(f"{cls.__name__}.from_dict: missing mandatory fields: {missing}")
        return from_dict(cls, data)

    @classmethod
    def from_json(cls: Type[QueryObjectType], file_name: str) -> QueryObjectType:
        with open(file_name) as f:
            return cls.from_dict(json.load(f))

```
Unfortunately, invoking the `from_json` method on JSON data obtained from the Kentik UI failed with this error:
```
Traceback (most recent call last):
  File "/Users/martin/projects/community_sdk_python/kentik_api_library/kentik_api/requests_payload/conversions.py", line 56, in from_dict
    return dacite.from_dict(data_class=data_class, data=data)
  File "/opt/homebrew/lib/python3.9/site-packages/dacite/core.py", line 63, in from_dict
    value = _build_value(type_=field.type, data=transformed_value, config=config)
  File "/opt/homebrew/lib/python3.9/site-packages/dacite/core.py", line 90, in _build_value
    return _build_value_for_collection(collection=type_, data=data, config=config)
  File "/opt/homebrew/lib/python3.9/site-packages/dacite/core.py", line 140, in _build_value_for_collection
    return data_type(_build_value(type_=item_type, data=item, config=config) for item in data)
  File "/opt/homebrew/lib/python3.9/site-packages/dacite/core.py", line 140, in <genexpr>
    return data_type(_build_value(type_=item_type, data=item, config=config) for item in data)
  File "/opt/homebrew/lib/python3.9/site-packages/dacite/core.py", line 92, in _build_value
    return from_dict(data_class=type_, data=data, config=config)
  File "/opt/homebrew/lib/python3.9/site-packages/dacite/core.py", line 63, in from_dict
    value = _build_value(type_=field.type, data=transformed_value, config=config)
  File "/opt/homebrew/lib/python3.9/site-packages/dacite/core.py", line 92, in _build_value
    return from_dict(data_class=type_, data=data, config=config)
  File "/opt/homebrew/lib/python3.9/site-packages/dacite/core.py", line 68, in from_dict
    raise WrongTypeError(field_path=field.name, field_type=field.type, value=value)
dacite.exceptions.WrongTypeError: wrong value type for field "queries.query.dimension" - should be "typing.List[kentik_api.public.query_object.DimensionType]" instead of value "['i_device_id', 'InterfaceID_dst']" of type "list"
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3437, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-d2c2d4e7ee61>", line 33, in <module>
    queries = {f.stem: QueryObject.from_json(f.as_posix()) for f in sorted(wd.glob("q?.json"))}
  File "<ipython-input-2-d2c2d4e7ee61>", line 33, in <dictcomp>
    queries = {f.stem: QueryObject.from_json(f.as_posix()) for f in sorted(wd.glob("q?.json"))}
  File "/Users/martin/projects/community_sdk_python/kentik_api_library/kentik_api/public/query_object.py", line 328, in from_json
    return cls.from_dict(json.load(f))
  File "/Users/martin/projects/community_sdk_python/kentik_api_library/kentik_api/public/query_object.py", line 323, in from_dict
    return from_dict(cls, data)
  File "/Users/martin/projects/community_sdk_python/kentik_api_library/kentik_api/requests_payload/conversions.py", line 58, in from_dict
    raise DeserializationError(data_class.__name__, str(err)) from err
kentik_api.public.errors.DeserializationError: QueryObject: wrong value type for field "queries.query.dimension" - should be "typing.List[kentik_api.public.query_object.DimensionType]" instead of value "['i_device_id', 'InterfaceID_dst']" of type "list"
```
This made me believe that `dacite.from_dict` cannot properly construct embedded lists of class instances (although it should have all required information :-/) . That was the main reason why I've implemented custom `from_dict` methods. Checking for mandatory attributes was added in order to make errors less obscure.
Maybe I've given up too early. I welcome any hints, suggestions, patches. The ultimate goal is to be able to use JSON data provided by the UI and construct complete `QueryObject` instances.
Note that it is actually possible to use a dictionary constructed from JSON data instead of a `QueryObject` instance in the call to `api.query.data`. So, for example this works:
```
with open('query_data.json') as f:
    qd = json.load(f)
r = api.query.data(qd)
```
In the code above, `api` is previously created instance of KentikAPI and `qd`is plain Python dictionary (as opposed to a `QueryObject` instance. The query payload produced by this call is valid as is the response. I believe it is a side-effect of serialization of `QueryObject` to JSON when constructing request payload (the duck-typed nature of Python saves the day). I though prefer to construct proper `QueryObject` instances from the JSON data first, so they can be programmatically modified.

Reply:
Maybe it's worth to experiment with `cast` config option (https://github.com/konradhalas/dacite#casting)
I was thinking about:
```python
from_dict(data_class=cls, data=data, config=Config(cast=[Enum]))
```
The problem I see is that we have to convert to a list of enums, not an enum, so this will probably not work.

Reply:
Support for collections in cast has been added: https://github.com/konradhalas/dacite/commit/cb3229a8006167e301fa4010458521abc9421bf5

So maybe following would work:
```python
from_dict(data_class=cls, data=data, config=Config(cast=[List]))
```